# importing the libraries

In [ ]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
from  bs4 import BeautifulSoup 
import pandas as pd


# Storing the table of wikipedia page to dataframe

In [ ]:
import requests
import lxml.html as lh
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')


In [ ]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    
    col.append((name,[]))
    
print(col)

In [ ]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
   
    
    #If row is not of size 10, the //tr data is not from our table 
    if len(T)!=3:
        break
   
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

# replacing the newline character and displaying th dataframe

In [ ]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)
df = df.replace('\n','', regex=True)
df.head()

# printing the number of columns

In [ ]:
df= df.drop(180)
for col in df.columns:
    print (col)

# replacing the column names new line character

In [ ]:
df.columns=df.columns.str.replace('\n','')


In [ ]:
df.dtypes


# assigning a new dataframe where Not assigned borough is not present

In [ ]:
df_upd= df[df['Neighborhood'] !='']
df_upd.head(12)

# reset the index

In [ ]:
df_upd= df_upd.reset_index(drop=True)

# The data given in wiki already has been grouped per postal code, hence modifying the GUI part to show as its there in coursera assignment

In [ ]:
df_upd = df_upd.replace('/',',', regex=True)

In [ ]:
df_upd.head(12)

# Shape of the dataframe

In [ ]:
df_upd.shape

# Part 2 of the assignment begins here

In [ ]:
dfOfLatLong = pd.read_csv('http://cocl.us/Geospatial_data')

In [22]:

dfOfLatLong= dfOfLatLong.rename(columns={'Postal Code' : 'Postal code'})
dfOfLatLong

Postal code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

# The CSV file with Data is uploaded in Data frame . time to merge the two dataframes to form one dtaframe for data.

In [23]:
dfOfLatLong = pd.merge(df_upd, dfOfLatLong, on="Postal code")


In [ ]:
dfOfLatLong.head(30)

# Part 3 of the assignment begins

In [30]:
LATITUDE= 43.706748
LONGITUDE = -79.594054
CLIENT_ID = 'PJHVABWBETOR53I2N3GKF1XVZZANCEYSNY5HK2IZC2LNY1BB'
CLIENT_SECRET = 'O0JDJQJPH04WFFHQN05VVEKTEA3WDY3NCIDCTUWIWK4XLBWI'
VERSION = '20200427'
LIMIT = 30
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&limit={}'.format(CLIENT_ID,CLIENT_SECRET,LATITUDE,LONGITUDE,VERSION,LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=PJHVABWBETOR53I2N3GKF1XVZZANCEYSNY5HK2IZC2LNY1BB&client_secret=O0JDJQJPH04WFFHQN05VVEKTEA3WDY3NCIDCTUWIWK4XLBWI&ll=43.706748,-79.594054&v=20200427&limit=30'

# Importing libraries

In [31]:
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library
import requests

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [44]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ea6db47dd0f850022ea51ef'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'suggestedRadius': 3378,
  'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 182,
  'suggestedBounds': {'ne': {'lat': 43.736973762936266,
    'lng': -79.55154379114254},
   'sw': {'lat': 43.67489961948987, 'lng': -79.63519857285468}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ba50113f964a52018cf38e3',
       'name': 'Fortinos',
       'location': {'address': '330 Queens Plate Dr',
        'crossStreet': 'at Hwy 27',
        'lat': 43.721438,
        'lng': -79.596291,
        'labeledLatLngs': [{'label': 'display

# Parse the results to select any restaurant out of the 182 restaurants as mentioned in above results

In [43]:
res=results['response']['groups'][0]['items']
res[2]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4b9924aaf964a5205e6435e3',
  'name': 'Golden Asian',
  'location': {'address': '123 - #7 Rexdale Blvd.',
   'lat': 43.71124779631982,
   'lng': -79.56596849779808,
   'labeledLatLngs': [{'label': 'display',
     'lat': 43.71124779631982,
     'lng': -79.56596849779808}],
   'distance': 2314,
   'cc': 'CA',
   'city': 'Toronto',
   'state': 'ON',
   'country': 'Canada',
   'formattedAddress': ['123 - #7 Rexdale Blvd.', 'Toronto ON', 'Canada']},
  'categories': [{'id': '4bf58dd8d48988d142941735',
    'name': 'Asian Restaurant',
    'pluralName': 'Asian Restaurants',
    'shortName': 'Asian',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/asian_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-4b9924aaf964a5205e6435e3-2'}

# converting the above JSON to dataframe

In [50]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

    df2 = json_normalize(res) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in df2.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = df2.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head()

name          categories  \
0                               Fortinos       Grocery Store   
1                        Mandarin Buffet  Chinese Restaurant   
2                           Golden Asian    Asian Restaurant   
3  The Keg Steakhouse + Bar - Dixon Road          Restaurant   
4                        Tre Rose Bakery              Bakery   

                  address  cc       city country          crossStreet  \
0     330 Queens Plate Dr  CA  Etobicoke  Canada            at Hwy 27   
1     200 Queens Plate Dr  CA    Rexdale  Canada            at Hwy 27   
2  123 - #7 Rexdale Blvd.  CA    Toronto  Canada                  NaN   
3          927 Dixon Road  CA    Toronto  Canada                  NaN   
4     2098 Kipling avenue  CA    Toronto  Canada  Kipling and rexdale   

   distance                                   formattedAddress  \
0      1645  [330 Queens Plate Dr (at Hwy 27), Etobicoke ON...   
1      1515  [200 Queens Plate Dr (at Hwy 27), Rexdale ON M...   
2      2314       [123 - #7 Rexdale Blvd., Toronto ON, Canada]   
3      2288       [927 Dixon Road, Toronto ON M9W 1J8, Canada]   
4      2268  [2098 Kipling avenue (Kipling and rexdale), To...   

                                      labeledLatLngs        lat        lng  \
0  [{'label': 'display', 'lat': 43.721438, 'lng':...  43.721438 -79.596291   
1  [{'label': 'display', 'lat': 43.7203596, 'lng'...  43.720360 -79.594387   
2  [{'label': 'display', 'lat': 43.71124779631982...  43.711248 -79.565968   
3  [{'label': 'display', 'lat': 43.68632856455964...  43.686329 -79.597353   
4  [{'label': 'display', 'lat': 43.7164913472275,...  43.716491 -79.569288   

  neighborhood postalCode state                        id  
0          NaN    M9W 7J7    ON  4ba50113f964a52018cf38e3  
1          NaN    M9W 6Y9    ON  4b5f8a3df964a52062c029e3  
2          NaN        NaN    ON  4b9924aaf964a5205e6435e3  
3          NaN    M9W 1J8    ON  4ad7719cf964a520870a21e3  
4          NaN    M9W 4K5    ON  4dad993e6e81d745d744c5fb

# Plotting all the resturants in the neighborhood

In [57]:
venues_map = folium.Map(location=[LATITUDE, LONGITUDE], zoom_start=12) # generate map centred around the the location postal code I picked

# add a red circle marker to represent the Conrad Hotel
folium.features.CircleMarker(
    [LATITUDE, LONGITUDE],
    radius=10,
    color='red',
    popup='Neighborhood M9W',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add all the restaurants as green circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='purple',
        popup=label,
        fill = True,
        fill_color='purple',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map